File: cpp/secret-handshake/secret_handshake.cpp
#include "secret_handshake.h"

#include <algorithm>
#include <string>
#include <vector>

namespace secret_handshake
{
    std::vector<std::string> commands(int n)
    {
        std::vector<std::string> result;
        if (n & 1)
            result.push_back("wink");
        if (n & 2)
            result.push_back("double blink");
        if (n & 4)
            result.push_back("close your eyes");
        if (n & 8)
            result.push_back("jump");
        if (n & 16)
            std::reverse(result.begin(), result.end());
        return result;
    }

} // namespace secret_handshake

File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where
    import Data.Bits ((.&.), xor)

    handshake :: Int -> [String]
    handshake n
        | n .&. 16 == 16 = reverse $ handshake (n `xor` 16)
        | n .&. 8 == 8 = handshake (n `xor` 8) ++ ["jump"]
        | n .&. 4 == 4 = handshake (n `xor` 4) ++ ["close your eyes"]
        | n .&. 2 == 2 = handshake (n `xor` 2) ++ ["double blink"]
        | n .&. 1 == 1 = ["wink"]
        | otherwise = []
